src/ray/util/logging.h

```diff
@@ -25,97 +19,86 @@ namespace ray {
-                        << __FILE__ << ":" << __LINE__   \
-                        << " Check failed: " #condition " "
+#define RAY_CHECK(condition)                                                          \
+  (condition) ? RAY_IGNORE_EXPR(0) : ::ray::Voidify() &                               \
+                                         ::ray::RayLog(__FILE__, __LINE__, RAY_FATAL) \
+                                             << " Check failed: " #condition " "
 
 #ifdef NDEBUG
 
 #define RAY_DCHECK(condition) \
-  RAY_IGNORE_EXPR(condition)  \
-  while (false) ::ray::internal::NullLog()
+  RAY_IGNORE_EXPR(condition); \
+  while (false) ::ray::RayLogBase()
 
 #else
 
 #define RAY_DCHECK(condition) RAY_CHECK(condition)
 
 #endif  // NDEBUG
 
-namespace internal {
+// To make the logging lib plugable with other logging libs and make
+// the implementation unawared by the user, RayLog is only a declaration
+// which hide the implementation into logging.cc file.
+// In logging.cc, we can choose different log libs using different macros.
 
-class NullLog {
+// This is also a null log which does not output anything.
+class RayLogBase {
  public:
-  template <class T>
-  NullLog &operator<<(const T &t) {
+  virtual ~RayLogBase(){};
+
+  template <typename T>
+  RayLogBase &operator<<(const T &t) {
     RAY_IGNORE_EXPR(t);
     return *this;
   }
 };
 
-class CerrLog {
+class RayLog : public RayLogBase {
  public:
-  CerrLog(int severity)  // NOLINT(runtime/explicit)
-      : severity_(severity),
-        has_logged_(false) {}
-
-  virtual ~CerrLog() {
-    if (has_logged_) {
-      std::cerr << std::endl;
-    }
-    // This code is duplicated in the FatalLog class.
-    if (severity_ == RAY_FATAL) {
-#if defined(_EXECINFO_H) || !defined(_WIN32)
-      void *buffer[255];
-      const int calls = backtrace(buffer, sizeof(buffer) / sizeof(void *));
-      backtrace_symbols_fd(buffer, calls, 1);
-#endif
-      std::abort();
-    }
-  }
-
-  template <class T>
-  CerrLog &operator<<(const T &t) {
-    if (severity_ != RAY_DEBUG) {
-      has_logged_ = true;
-      std::cerr << t;
+  RayLog(const char *file_name, int line_number, int severity);
+  virtual ~RayLog();
+
+  template <typename T>
+  RayLogBase &operator<<(const T &t) {
+    if (implement == nullptr) {
+      RAY_IGNORE_EXPR(t);
+    } else {
+      this->Stream() << t;
     }
     return *this;
   }
 
- protected:
-  const int severity_;
-  bool has_logged_;
+  // The init function of ray log for a program which should be called only once.
+  // If logDir is empty, the log won't output to file.
+  static void StartRayLog(const std::string &appName, int severity_threshold = RAY_ERROR,
+                          const std::string &logDir = "");
+  // The shutdown function of ray log which should be used with StartRayLog as a pair.
+  static void ShutDownRayLog();
+
+ private:
+  std::ostream &Stream();
+  const char *file_name_;
+  int line_number_;
+  int severity_;
+  void *implement;
```

Comment:
maybe logging_provider_? (up to you, but it should have an underscore at the end)

Reply:
Instead of a void* pointer and the explicit deletes, we can also make
RayLog a template (either RayLog<CerrLog> or RayLog<google::LogMessage>) and store this in a unique pointer (call .reset to allocate a new one).

Reply:
I use `std::unique_ptr` but not in the template way. I use `typedef` to define type `LoggingProvider` equals to `google::LogMessage` or `ray::CerrLog` and declare `std::unique_ptr<LoggingProvider> logging_provider_;` in the class.